relayer/src/chain/cosmos.rs

```diff
@@ -298,72 +221,130 @@ impl Chain for CosmosSDKChain {
+
+        Ok((Box::new(lc), Some(supervisor_thread)))
     }
 
-    fn build_consensus_state(
+    fn init_event_monitor(
         &self,
-        light_block: Self::LightBlock,
-    ) -> Result<Self::ConsensusState, Error> {
-        Ok(TMConsensusState::from(light_block.signed_header.header))
+        rt: Arc<Mutex<TokioRuntime>>,
+    ) -> Result<
+        (
+            channel::Receiver<EventBatch>,
+            Option<thread::JoinHandle<()>>,
+        ),
+        Error,
+    > {
+        let (mut event_monitor, event_receiver) =
+            EventMonitor::new(self.config.id.clone(), self.config.rpc_addr.clone(), rt)?;
+
+        event_monitor.subscribe().unwrap();
+        let monitor_thread = thread::spawn(move || event_monitor.run());
+
+        Ok((event_receiver, Some(monitor_thread)))
     }
 
-    fn build_header(
-        &self,
-        trusted_light_block: Self::LightBlock,
-        target_light_block: Self::LightBlock,
-    ) -> Result<Self::Header, Error> {
-        let trusted_height =
-            ICSHeight::new(self.id().version(), trusted_light_block.height().into());
+    fn id(&self) -> &ChainId {
+        &self.config().id
+    }
 
-        Ok(TMHeader {
-            trusted_height,
-            signed_header: target_light_block.signed_header.clone(),
-            validator_set: target_light_block.validators,
-            trusted_validator_set: trusted_light_block.validators,
-        })
+    fn keybase(&self) -> &KeyRing {
+        &self.keybase
+    }
+
+    fn query(&self, data: Path, height: ICSHeight, prove: bool) -> Result<QueryResponse, Error> {
+        let path = TendermintABCIPath::from_str(IBC_QUERY_PATH).unwrap();
+
+        let height =
+            Height::try_from(height.version_height).map_err(|e| Kind::InvalidHeight.context(e))?;
+
+        if !data.is_provable() & prove {
+            return Err(Kind::Store
+                .context("requested proof for a path in the privateStore")
+                .into());
+        }
+
+        let response =
+            self.block_on(abci_query(&self, path, data.to_string(), height, prove))??;
+
+        // TODO - Verify response proof, if requested.
+        if prove {}
+
+        Ok(response)
+    }
+
+    /// Send one or more transactions that include all the specified messages
+    fn send_msgs(&mut self, proto_msgs: Vec<Any>) -> Result<Vec<String>, Error> {
+        if proto_msgs.is_empty() {
+            return Ok(vec!["No messages to send".to_string()]);
+        }
+        let mut res = vec![];
+
+        let mut n = 0;
+        let mut size = 0;
+        let mut msg_batch = vec![];
+        for msg in proto_msgs.iter() {
+            msg_batch.append(&mut vec![msg.clone()]);
+            let mut buf = Vec::new();
+            prost::Message::encode(msg, &mut buf).unwrap();
+            n += 1;
+            size += buf.len();
+            if n >= self.max_tx_num() || size >= self.max_tx_size() {
```

Comment:
Just to be clear: a transaction as we submit it to `broadcast_tx_commit` will not really reflect the `max_tx_size` bound. The transcation may be slightly bigger. This is because we do additional wrapping and add additional fields inside `send_tx`, prior to `broadcast_tx_commit`. But I suppose this is not an issue (at the moment, at least).

Reply:
Yes, it's close to that. We need to do some work here, these changes are temporary until we get to it.